Add native pipeline support and texture related functionality

sepcnt opened the pull request:
This PR addresses bounty https://github.com/MonoGame/MonoGame/issues/8641 by replacing the `FreeImage` and `StbImageSharp` dependencies with a new, in-house native C++ pipeline library (`mgpipeline`) for texture importing.
The changes improve modularity and performance by consolidating image-related operations into the `mgpipeline` library. Additionally, submodule management and project dependencies have been updated to reflect these changes.

### Image Processing Refactor:
* [`MonoGame.Framework.Content.Pipeline/Graphics/GraphicsUtil.cs`](diffhunk://#diff-37a0beedf17141b267965f66061702d42451da617536f0244474d6cf907261b2L20-R94): Refactored the `Resize` method to use the `MGCP_Bitmap` structure and `MGCP.MP_ResizeBitmap` function from the `mgpipeline` library, replacing FreeImage-based resizing logic.
* [`MonoGame.Framework.Content.Pipeline/TextureImporter.cs`](diffhunk://#diff-d274e21bce6bb79d1b085fcf3355b1595041ecee26848bce4b085b9c5551d823L77-L209): Replaced FreeImage and Stb-based image importing with `MGCP.MP_ImportBitmap` and added support for new texture types. Removed redundant methods for channel swapping and BMP loading.

### Native Library Integration:
* [`MonoGame.Framework.Content.Pipeline/Native/Texture.Interop.cs`](diffhunk://#diff-7b1c87b82a48fd442dacc69b246a854fa6934758fcb75388dfee5e8780e94276R1-R52): Introduced a new interop file defining `MGCP_Bitmap` structure and native methods (`MP_ImportBitmap`, `MP_ResizeBitmap`, `MP_ExportBitmap`, etc.) for interacting with the `mgpipeline` library.
* [`MonoGame.Framework.Content.Pipeline/MonoGame.Framework.Content.Pipeline.csproj`](diffhunk://#diff-393aad48627ab537c1c1e207c5ecc909589fb716844dfaf82fa3ce5739316ef1R93-R104): Added `mgpipeline` native library files (`mgpipeline.dll`, `libmgpipeline.so`, `libmgpipeline.dylib`) to the project for cross-platform support. Removed references to FreeImage and Stb libraries.

### Submodule and Workflow Updates:
* [`.gitmodules`](diffhunk://#diff-fe7afb5c9c916e521401d3fcfb4277d5071798c3baf83baf11d6071742823584R28-R30): Added a new submodule for `external/stb`, ensuring proper version control for dependencies.
* [`.github/workflows/main.yml`](diffhunk://#diff-7829468e86c1cc5d5133195b5cb48e1ff6c75e3e9203777f6b2e379d9e4882b3R131-R132): Updated the repository cloning step to include submodules recursively in deploy stage.

### File Format Export:
* [`MonoGame.Framework.Content.Pipeline/Utilities/FileHelpers.cs`](diffhunk://#diff-1f025d7f4005e47796cc286813c602fa8af27475b50a095469cfe8af7011eee0L48-R82): Refactored PNG export logic to use `MGCP.MP_ExportBitmap` instead of StbImageWriteSharp, improving error handling and efficiency.

### Cleanup and Simplification:
* [`MonoGame.Framework.Content.Pipeline/MonoGame.Framework.Content.Pipeline.csproj`](diffhunk://#diff-393aad48627ab537c1c1e207c5ecc909589fb716844dfaf82fa3ce5739316ef1L52-L54): Removed unused references to StbImageSharp and StbImageWriteSharp, as well as the FreeImage library. [[1]](diffhunk://#diff-393aad48627ab537c1c1e207c5ecc909589fb716844dfaf82fa3ce5739316ef1L52-L54) [[2]](diffhunk://#diff-393aad48627ab537c1c1e207c5ecc909589fb716844dfaf82fa3ce5739316ef1L78)
* [`MonoGame.Framework.Content.Pipeline/Graphics/GraphicsUtil.cs`](diffhunk://#diff-37a0beedf17141b267965f66061702d42451da617536f0244474d6cf907261b2L52-L61): Removed the `BGRAtoRGBA` method, as channel swapping is no longer required with the new library.

Commit:
Add native pipeline support and texture importing functionality

- Updated .gitignore to include Premake generated files.
- Added new submodule for stb_image.
- Created native pipeline library with C API for bitmap import.
- Implemented texture importing using native methods in TextureImporter.
- Enhanced build process for native pipeline with Premake and MSBuild.
- Added README and API headers for the native pipeline.

harry-cpp commented on native/pipeline/mgcp_pipeline.cpp:
```diff
@@ -0,0 +1,58 @@
+        return NULL;
+
+    MGCP_Bitmap* bmp = (MGCP_Bitmap*)malloc(sizeof(MGCP_Bitmap));
+    if (!bmp) {
```
Braces should be in new lines.

harry-cpp commented on native/pipeline/mgcp_pipeline.cpp:
```diff
@@ -0,0 +1,58 @@
+    if (is_16_bit)
+        data = stbi_load_from_file_16(f, &width, &height, &channels, 4);
+    else
+		data = stbi_load_from_file(f, &width, &height, &channels, 4);
```
In a lot of places in code there are tabs instead of spaces, ensure that spaces are used.

harry-cpp commented on MonoGame.Framework.Content.Pipeline/MonoGame.Framework.Content.Pipeline.csproj:
```diff
@@ -94,6 +94,9 @@
       <PackagePath>runtimes\win-x64\native</PackagePath>
       <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
     </Content>
+	<Content Include="..\Artifacts\monogame.native.pipeline\windows\$(Configuration)\monogame.native.pipeline.dll">
```
Tabs instead of spaces here.

harry-cpp commented on build/BuildToolsTasks/BuildContentPipelineTask.cs:
```diff
@@ -7,6 +7,21 @@ public sealed class BuildContentPipelineTask : FrostingTask<BuildContext>
+        // Build the native pipeline library.
+        exit = context.StartProcess("msbuild", new ProcessSettings { WorkingDirectory = "native/pipeline", Arguments = "pipeline.sln /p:Configuration=Release /p:Platform=x64" });
+        if (exit != 0)
+            throw new Exception($"Native Pipeline build failed! {exit}");
```
You need to build the library for macOS and Linux as well.

harry-cpp commented on native/pipeline/README.md:
```diff
@@ -0,0 +1,8 @@
+
+## Structure
+- All C++ source code for the native pipeline will live here.
+- The library will be built as `monogame.native.pipeline` (dll/so/dylib).
```
Library name should be `mgpipeline`

harry-cpp commented on MonoGame.Framework.Content.Pipeline/TextureImporter.cs:
```diff
@@ -74,139 +72,44 @@ public override TextureContent Import(string filename, ContentImporterContext co
-                case FREE_IMAGE_TYPE.FIT_RGBAF:
-                    face = new PixelBitmapContent<Vector4>(width, height);
-                    break;
+                MGCP_Bitmap* nativeBitmap = MGCP.MP_ImportBitmap(filename);
```
Instead of returning a pointer and then having to free it, set it up so that the structure gets allocated on the C# side and the C++ side just fills it up, ie:

```cs
MGCP_Bitmap nativeBitmap;
MGCP.MP_ImportBitmap(filename, ref nativeBitmap);
```

There should also be an error handling system like `SDL_GetError` that can help if an error occurs to give us a proper error message of where it happened.

harry-cpp commented on native/pipeline/mgcp_pipeline.cpp:
```diff
@@ -0,0 +1,58 @@
+{
+    FILE* f = stbi__fopen(filename, "rb");
+    int is_16_bit;
+    if (!f) return NULL;
```
Use `nullptr` instead of `NULL`

harry-cpp commented on native/pipeline/mgcp_pipeline.cpp:
```diff
@@ -0,0 +1,58 @@
+#define STB_IMAGE_IMPLEMENTATION
```
This define should be passed from the compiler if its needed for stb.

harry-cpp reviewed (changes requested):
You can remove the test that is failing, for the switching to stb_image we have decided to not support any image format not supported by it for importing.

Commit:
Overall fix for native pipeline

- Rename native pipeline to `mgpipeline`
- Remove TIFF format test and asset
- Update .gitignore and enhance BuildContentPipelineTask for cross-platform support
- Remove system-specific path from platform target in premake5.lua
- Refactor API export macros for improved cross-platform compatibility
- Refactor `MGCP_Bitmap` to use reference types for better memory management and error handling
- Fix Linux native pipeline library reference in project file
- Refactor `MP_ImportBitmap` to satisfy GCC

Commit:
Add HDR support in native pipeline `MP_ImportBitmap`

sepcnt commented:
To totally remove FreeImage and StbImage(Write)Sharp, there are few more APIs to implement:
- `MG_ResizeBitmap`
- `MG_ExportBitmap`

However, there's a critical bug in `BasicUHelpers` and `CrunchHelpers`:
https://github.com/MonoGame/MonoGame/blob/7c11dd086ec695016cdac7f8a37f5448e1e3fb36/MonoGame.Framework.Content.Pipeline/Utilities/BasisUHelpers.cs#L375-L377

https://github.com/MonoGame/MonoGame/blob/7c11dd086ec695016cdac7f8a37f5448e1e3fb36/MonoGame.Framework.Content.Pipeline/Utilities/FileHelpers.cs#L51

That means all textures are converts to rgba8 regardless of there original color depth (16bit or float)

harry-cpp commented on MonoGame.Framework.Content.Pipeline/MonoGame.Framework.Content.Pipeline.csproj:
```diff
@@ -94,6 +94,15 @@
       <PackagePath>runtimes\win-x64\native</PackagePath>
       <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
     </Content>
+    <Content Include="..\Artifacts\monogame.native.pipeline\$(Configuration)\mgpipeline.dll" Condition=" '$(OS)' == 'Windows_NT' ">
```
Use file exists instead of system check.

harry-cpp commented on native/pipeline/mgcp_pipeline.cpp:
```diff
@@ -0,0 +1,50 @@
+
+    if (stbi_is_hdr_from_file(f))
+    {
+		bitmap.format = MGTextureFormat::RgbaF;
```
tabs in file.

harry-cpp commented on native/pipeline/premake5.lua:
```diff
@@ -0,0 +1,37 @@
+-- file 'LICENSE.txt', which is part of this source code package.
+
+function pipeline_native()
+    platform_target_path = "../../Artifacts/monogame.native.pipeline/%{cfg.buildcfg}"
```
`monogame.native.pipeline` -> `mgpipelin`

harry-cpp commented on MonoGame.Framework.Content.Pipeline/TextureImporter.cs:
```diff
@@ -74,139 +72,55 @@ public override TextureContent Import(string filename, ContentImporterContext co
-            FreeImage.ConvertToRawBits(bytes, fBitmap, pitch, bpp, redMask, greenMask, blueMask, true);
-            // Create the Pixel bitmap content depending on the image type
-            switch(imageType)
+            unsafe
```
No need for unsafe now.

harry-cpp commented:
Since its happening on the current develop branch I'd recommend just leaving it working as is in here, and just opening a new issue about it (with a couple of comments in code about the issue).

Tho if its something thats easy to fix, feel free to do it here.

sepcnt commented:
> Since its happening on the current develop branch I'd recommend just leaving it working as is in here, and just opening a new issue about it (with a couple of comments in code about the issue).
> 
> Tho if its something thats easy to fix, feel free to do it here.

Okay, it's weird but works. I opened an issue for it here https://github.com/MonoGame/MonoGame/issues/8874

Commit:
Remove duplicate `stb` headers in native/monogame/common

harry-cpp reviewed (changes requested):
You'll need to setup a way to package up MonoGame.Content.Pipeline for all the platforms, right now each platform is makings its own nuget that only has 1 native assembly in it for itself which would break stuff.

To do this I recommend:
- Upload each mgpipeline native lib as an artifact during build
- Download each mgpipeline native lib during Deploy, place them in the Artifacts folder and combine then rebuild just the MonoGame.Content.Pipeline by creating a brand new cake task

We use GitHub actions located in `.github/workflows` and C# Cake project located by opening `Build.sln` in root for CI operations, I believe you just need to modify the C# side of things for this, you can test on your local fork by enabling github actions on it.

Another thing to note is that `mgpipeline` needs to be statically linked and the CI should check for that. You can grab the code for that from: <https://github.com/MonoGame/MonoGame.Library.BuildScripts/blob/5cd1ee39dfa0e570b125260d4d19753232c1c29e/Tasks/TestLinuxTask.cs#L5> (there is 1 task for each platform)

harry-cpp commented:
PS. Feel free to ping me on discord for me to review the PR or if you have any questions about this, I should notice it much quicker then on github ^^

Commit:
Refactor native texture handling: replace FreeImage with custom bitmap processing, add resizing functionality, and update related enums and structures.

Commit:
Refactor bitmap handling: remove stb image dependencies, implement custom export functionality.

Commit:
Refactor bitmap handling: replace unsafe code with GCHandle for memory management, update import/export paths.

harry-cpp commented on build/BuildToolsTasks/BuildContentPipelineTask.cs:
```diff
@@ -7,6 +7,30 @@ public sealed class BuildContentPipelineTask : FrostingTask<BuildContext>
+            if (exit != 0)
+                throw new Exception($"Native Pipeline Premake generation failed! {exit}");
+
+            exit = context.StartProcess("make", new ProcessSettings { WorkingDirectory = "native/pipeline", Arguments = "config=release" });
```
Forgotten exit code check.

harry-cpp commented on MonoGame.Framework.Content.Pipeline/Graphics/GraphicsUtil.cs:
```diff
@@ -17,29 +18,80 @@ internal static BitmapContent Resize(this BitmapContent bitmap, int newWidth, in
+                IntPtr err = MGCP.MP_ResizeBitmap(ref srcBitmap, ref dstBitmap);
+                if (err != IntPtr.Zero)
+                {
+                    string errorMsg = System.Runtime.InteropServices.Marshal.PtrToStringAnsi(err);
```
Use `PtrToStringUTF8` (same for the other occurrence).

Commit:
Update error message handling and native pipeline on Windows to use UTF-8 encoding.

sepcnt committed:
Merge branch 'develop' into develop

Commit:
Pack all-in-one nuget package for ContentPipeline during the CI

Commit:
Merge branch 'develop' of https://github.com/sepcnt/MonoGame into develop

Commit:
Fix CI and build configuration

- Fix bpp calculation bug to avoid wasting memory
- Add static library check task
- Fix premake5 configuration for macOS universal support, static linkage on
  Windows

tomspilman commented on MonoGame.Framework.Content.Pipeline/Graphics/GraphicsUtil.cs:
```diff
@@ -17,29 +18,80 @@ internal static BitmapContent Resize(this BitmapContent bitmap, int newWidth, in
+                    throw new InvalidContentException($"Bitmap resize failed: {errorMsg}");
+                }
+
+                Marshal.Copy(dstBitmap.data, newBytes, 0, newBytes.Length);
```
@harry-cpp I wonder if we should have the C# side responsible for the memory of both the source and dest bitmaps passed to `MGCP.MP_ResizeBitmap`.  Then we potentially don't need this extra copy here?

tomspilman commented on build/StaticLibCheck.cs:
```diff
@@ -0,0 +1,207 @@
+using Cake.Common.Tools.VSWhere.Latest;
+namespace BuildScripts;
+
+public class StaticLibCheckTask
```
A comment here stating the purpose of this class and how it works will help future development.

harry-cpp replied:
Also `StaticLibCheckTask` -> `StaticLibCheck` for the name.

tomspilman commented on native/monogame/include/api_common.h:
```diff
@@ -19,8 +19,30 @@ typedef double               mgdouble;
 
 #define MG_FIELD_OFFSET(off, type, name) struct { char _pad_##name[off]; type name; }
 
+#if defined(_WIN32) || defined(__CYGWIN__)
```
Not sure how these changes in the export will affect the consoles... will have to test it later and fix things.  It however worked before.

Commit:
Manage MP_ResizeBitmap dstBitmap.data on C# side and update StaticLibCheck class for clarity

harry-cpp commented:
Looking good so merging! Thank you very much for getting this work done @sepcnt !
